File: wheels/7.refersh/7.refersh/ViewController.m
//
//  ViewController.m
//  7.Refersh
//
//  Created by Hayder on 2016/10/31.
//  Copyright © 2016年 wangzhenhai. All rights reserved.
//

#import "ViewController.h"
#import "ZHRefersh.h"

@interface ViewController ()

@property (nonatomic, strong) NSMutableArray *dataList;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    //注册cell
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"cell"];
    
    self.tableView.tableFooterView = [[UIView alloc] init];
    
    __weak typeof(self) weakSwlf = self;
    [self.tableView.footerView setFooterRefersh:^{

        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            
            [weakSwlf.tableView.footerView endRefershing];
            
            for (int i=0; i< 15; i++) {
                
                [weakSwlf.dataList addObject:[NSString stringWithFormat:@"添加数据%d",i]];
            }
            
            [weakSwlf.tableView reloadData];
            
            
        });
    }];
    
    [self.tableView.headerView setHeaderRefersh:^{
       
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            
            [weakSwlf.tableView.headerView endRefershing];
            
            for (int i=0; i< 15; i++) {
                
                [weakSwlf.dataList insertObject:[NSString stringWithFormat:@"插入数据%d",i] atIndex:0];
            }
            
            [weakSwlf.tableView reloadData];
            
            
        });
    }];
   
}


#pragma mark --------------------tableViewDelegate&&dataSource------------------------
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.dataList.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
    
    cell.textLabel.text = self.dataList[indexPath.row];
    return cell;
}

#pragma mark ------------------------setter && getter------------------------
- (NSMutableArray *)dataList
{
    if (!_dataList) {
        
        _dataList = [NSMutableArray array];
        
        for (int i=0; i< 15; i++) {
            
            [_dataList addObject:[NSString stringWithFormat:@"原始数据%d",i]];
        }
    }
    
    return _dataList;
}

- (void)dealloc
{
    NSLog(@"我被释放了");
}
@end

File: Demo/Swift/语法/10.Swift3.0 -tableview/10.Swift3.0 -tableview/ViewController.swift
//
//  ViewController.swift
//  10.Swift3.0 -tableview
//
//  Created by Hayder on 16/9/26.
//  Copyright © 2016年 wangzhenhai. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UITableViewDataSource {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        setUp()
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 20
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "cellID", for: indexPath)
        //提示： textLabel 是可选的
        //如果 textLabel 有，就使用，如果没有，就什么也不做
        cell.textLabel?.text = "\(indexPath.row)"
        
        return cell
        
    }
    
    func setUp() {
        
        //1.创建表格
        let tv = UITableView(frame: view.bounds, style: .plain)
        
        //2.添加到视图
        view.addSubview(tv)
        
        //3.注册可重用项
        tv.register(UITableViewCell.self, forCellReuseIdentifier: "cellID")
        
        //4.设置数据源
        tv.dataSource = self;
    }
}

